### ISSUE TITLE
SIM115 triggers when an ExitStack is passed into a function

### ISSUE THREAD

**Issue by tjstum** [astral-sh/ruff#14110]
I am using ruff 0.7.2. I searched for  SIM115 and `ExitStack` before filing this issue (and found #1945).

If you have code like this:
```python
from contextlib import ExitStack

with ExitStack() as stack:
    stack.enter_context(open("/path/to/file"))
```
SIM115 does not trigger. However, if the stack object is passed into the function (correctly annotated), the message is issued:
```python
from contextlib import ExitStack

def open_the_file_elsewhere(exit_stack: ExitStack) -> None:
    exit_stack.enter_context(open("/path/to/file"))
```

I'm not sure to what extent ruff follows annotated function arguments, but I would expect this pattern not to trigger the message.

**Comment by AlexWaygood** [comment]
Thanks, I agree that this is definitely something we _should_ support. However, it won't be possible until we have more advanced type inference. This is something we're actively working on, but it will be a while before it arrives, I'm afraid.

**Comment by cdleonard** [comment]
I hit the same issue, using exit_stack as a pytest fixture.

Not clear why fixing this needs advanced type inference. I would be happy if it was SIM115 would be prevented if you have an argument that is annotated as an ExitStack.
